msbuild/Xamarin.MacDev.Tasks/Tasks/AnalyzeFileChanges.cs

```diff
@@ -0,0 +1,97 @@
+
+		[Required]
+		public ITaskItem HashesFile { get; set; }
+
+		[Output]
+		public ITaskItem [] ChangedFiles { get; set; } = [];
+
+		public IEnumerable<ITaskItem> GetAdditionalItemsToBeCopied () => [];
+
+		//We need the HashesFile input to be copied to the Mac. Since it's the only ITaskItem input, it's safe to return true
+		public bool ShouldCopyToBuildServer (ITaskItem item) => true;
+
+		//In case it's a remote execution, we don't want empty output files to be copied since we need the real files to be copied
+		public bool ShouldCreateOutputFile (ITaskItem item) => false;
+
+		public override bool Execute ()
+		{
+			if (ShouldExecuteRemotely ()) {
+				return new TaskRunner (SessionId, BuildEngine4).RunAsync (this).Result;
+			}
+
+			if (!Directory.Exists (WorkingDirectory)) {
+				Log.LogError ("Unable to analyze file changes in '{0}'. Directory not found.", WorkingDirectory);
+
+				return false;
+			}
+
+			if (!File.Exists (HashesFile.ItemSpec)) {
+				Log.LogError ("Unable to analyze file changes in '{0}'. The hashes file '{1}' does not exist", WorkingDirectory, HashesFile.ItemSpec);
+
+				return false;
+			}
+
+			try {
+				var changedFiles = new List<ITaskItem> ();
+				//Method from Xamarin.Messaging.Common
+				using HashAlgorithm hashAlgorithm = Hash.GetAlgorithm ();
+				//Gets a dictionary of file names and hashes from the HashesFile
+				IDictionary<string, string> fileHashesList = GetFileHashesList ();
+
+				foreach (string file in Directory.GetFiles (WorkingDirectory, "*", SearchOption.TopDirectoryOnly)) {
+					//We only want the files that are present in the HashesFile
+					if (fileHashesList.TryGetValue (Path.GetFileName (file), out string localHash)) {
+						using var stream = new FileStream (file, FileMode.Open, FileAccess.Read);
+						//Extension method from Xamarin.Messaging.Common
+						string remoteHash = hashAlgorithm.ComputeHashAsString (stream);
+
+						//If the hash from the hashes file (local hash) is different than the calculated hash of the file, it means local and remote versions are different
+						if (!string.Equals (localHash, remoteHash, StringComparison.OrdinalIgnoreCase)) {
+							changedFiles.Add (new TaskItem (file));
+						}
+					}
+				}
+
+				ChangedFiles = changedFiles.ToArray ();
+
+				return true;
+			} catch (Exception ex) {
+				Log.LogError ("Unable to analyze file changes in '{0}'. An unexpected error occurred.", WorkingDirectory);
+				Log.LogErrorFromException (ex);
+
+				return false;
+			}
+		}
+
+		IDictionary<string, string> GetFileHashesList ()
+		{
+			var fileHashesList = new Dictionary<string, string> ();
+
+			//Expected format of the hashes file lines: Foo.dll>c4a84cdd787cb31aaa46e8282f7d288f0641fdaa4252ac78979340131c8b9110
+			foreach (var line in File.ReadLines (HashesFile.ItemSpec).Where (l => l.Contains (">"))) {
+				string fileName = line.Substring (0, line.IndexOf (">"));
+				string fileHash = line.Substring (line.IndexOf (">") + 1);
```

Comment:
Each line is searched three times for the separator here. This is faster:

```suggestion
			foreach (var line in File.ReadLines (HashesFile.ItemSpec)) {
				var idx = line.IndexOf ('>');
				if (idx == -1)
					continue;
				string fileName = line.Substring (0, idx);
				string fileHash = line.Substring (idx + 1);
```

Reply:
Done